core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2049,12 +2054,45 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
     * res0: Unit = ()
     * }}}
     */
-  def parEvalMapUnordered[F2[x] >: F[
-    x
-  ]: Concurrent, O2](
+  def parEvalMapUnordered[F2[x] >: F[x]: Concurrent, O2](
       maxConcurrent: Int
   )(f: O => F2[O2]): Stream[F2, O2] =
-    map(o => Stream.eval(f(o))).parJoin(maxConcurrent)
+    Stream
+      .eval {
+        (
+          Semaphore[F2](maxConcurrent),
+          Queue.bounded[F2, O2](1),
+          Deferred[F2, Unit],
+          Deferred[F2, Either[Throwable, Unit]]
+        ).tupled
+      }
+      .flatMap { case (semaphore, queue, seenNone, halter) =>
+        val completeQueue =
+          seenNone.tryGet
+            .flatMap(_.fold(false.pure[F2])(_ => semaphore.available.map(_ == maxConcurrent)))
+            .flatMap(bool => if (bool) halter.complete(().asRight).void else ().pure[F2])
+
+        val pullExecAndOutput =
+          noneTerminate
```

Comment:
I think there shouldn't be an `interruptWhen`, because concurrently does this for us. However I wanted to draw an attention to this line in that sense.